### DISCUSSION TITLE
Built client and stdlib import 20 missing files

### DISCUSSION THREAD

**Discussion by garthk** [observablehq/framework#1082] [Q&A]
G'day! In an attempt to further bundle the output to get down to one HTML file if possible, I ran into some strife.  With this Rollup config…

```js
import { rollupPluginHTML as html } from '@web/rollup-plugin-html';

export default {
  input: 'dist/index.html',
  output: { dir: 'bundledist' },
  plugins: [html({ extractAssets: true, })],
};
```

… and this command…

```bash
node_modules/.bin/rollup --config rollup.config.js
```

… I get one of 20 errors resembling:

```plain
[!] RollupError: Could not resolve "../_npm/apache-arrow@15.0.1/+esm.js" from "dist/_observablehq/stdlib.js"
```

I figure the import is unreachable: the built page seems to work. But, it's blocking Rollup here, and there might be some underlying problem you need to know about that results in:

- Your output containing your code that tries to import third party code that's missing
- Your output missing third party code despite it being imported by your code
 
Looking for all such imports:

```plain
$ find dist -type f -exec grep -Eoh 'import\("[^"]+"\)' {} + | sort | uniq
import("../_npm/@duckdb/duckdb-wasm@1.28.0/+esm.js")
import("../_npm/@observablehq/plot@0.6.14/+esm.js")
import("../_npm/apache-arrow@15.0.1/+esm.js")
import("../_npm/arquero@5.4.0/+esm.js")
import("../_npm/d3-dsv@3.0.1/+esm.js")
import("../_npm/d3@7.9.0/+esm.js")
import("../_npm/echarts@5.5.0/dist/echarts.esm.min.js")
import("../_npm/htl@0.3.1/+esm.js")
import("../_npm/leaflet@1.9.4/+esm.js")
import("../_npm/lodash@4.17.21/+esm.js")
import("../_npm/mapbox-gl@3.2.0/+esm.js")
import("../_npm/parquet-wasm@0.5.0/esm/arrow1.js")
import("../_npm/topojson-client@3.1.0/+esm.js")
import("./stdlib/dot.js")
import("./stdlib/duckdb.js")
import("./stdlib/inputs.js")
import("./stdlib/mermaid.js")
import("./stdlib/sqlite.js")
import("./stdlib/tex.js")
import("./stdlib/vega-lite.js")
import("./stdlib/vgplot.js")
import("./stdlib/xlsx.js")
import("./stdlib/zip.js")
```

Of these, three are present in the output:

```plain
# find dist -type f -exec grep -Eoh 'import\("[^"]+"\)' {} + | sort | uniq | cut -f 2 -d '"' | while IFS= read -r dep; do if [[ -f dist/_observablehq/$dep ]]; then echo "$dep"; fi; done
../_npm/@observablehq/plot@0.6.14/+esm.js
../_npm/d3-dsv@3.0.1/+esm.js
../_npm/d3@7.9.0/+esm.js
```

Adding the others to the Rollup config as `external` lets Rollup succeed, but it's a bit fiddly.

**Comment by mbostock** [comment]
We don’t support this use case, so I can’t provide too much support here. But I will say that this is expected in that `observablehq:stdlib` references a number of third-party libraries (so-called [recommended libraries](https://observablehq.com/framework/javascript/imports#implicit-imports)) for implicit imports; these libraries are only imported (and included in `_npm`) if they are actually used, as when you reference `d3` or `Arrow` from Markdown. Therefore if you use Rollup to bundle the code, you probably want to mark everything in `_npm` as external; otherwise you’ll have to know what your project actually references, and only mark as external libraries you’re not use.
